binary_search_tree/tree.py

```diff
@@ -1,59 +1,152 @@
+        if new_node.key <= current.key:
+            current.left = self.add_helper(current.left, new_node)
+        elif new_node.key > current.key:
+            current.right = self.add_helper(current.right, new_node)
+
+        return current
+
+    # Time Complexity: O(log n) because we cut our search in half each pass through
+    # Space Complexity:  O(1)
     def find(self, key):
-        pass
+        current = self.root
+        if current is None:
+            return None
 
-    # Time Complexity: 
-    # Space Complexity: 
+        while current != None:
+            if key == current.key:
+                return current.value
+            elif key < current.key:
+                current = current.left
+            elif key > current.key:
+                current = current.right
+
+    # Time Complexity: O(n) since we visit every node to collect the value of it and append it
+    # Space Complexity:O(n) since we create & increase the size of the array we return
     def inorder(self):
-        pass
+        current = self.root
+        arr = []
 
-    # Time Complexity: 
-    # Space Complexity:     
+        if current is None:
+            return arr
+
+        return self.inorder_helper(current, arr)
+
+    def inorder_helper(self, current, arr):
+        if current is None:
+            return arr
+
+        current.left = self.inorder_helper(current.left, arr)
+        arr.append({"key": current.key, "value": current.value})
+        current.right = self.inorder_helper(current.right, arr)
+
+        return arr
+
+    # Time Complexity: O(n)
+    # Space Complexity:O(n)
     def preorder(self):
-        pass
+        current = self.root
+        arr = []
+
+        if current is None:
+            return arr
+
+        return self.preorder_helper(current, arr)
+
+    def preorder_helper(self, current, arr):
+        if current is None:
+            return arr
 
-    # Time Complexity: 
-    # Space Complexity:     
+        arr.append({"key": current.key, "value": current.value})
+        current.left = self.preorder_helper(current.left, arr)
+        current.right = self.preorder_helper(current.right, arr)
+
+        return arr
+
+    # Time Complexity: O(n)
+    # Space Complexity:O(n)
     def postorder(self):
-        pass
+        current = self.root
+        arr = []
+
+        if current is None:
+            return arr
+
+        return self.postorder_helper(current, arr)
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def postorder_helper(self, current, arr):
+        if current is None:
+            return arr
+
+        current.left = self.postorder_helper(current.left, arr)
+        current.right = self.postorder_helper(current.right, arr)
+        arr.append({"key": current.key, "value": current.value})
+
+        return arr
+
+    # Time Complexity: O(n) since we visit every node
+    # Space Complexity:O(1) since we are not creating or modifying any data structures
     def height(self):
```

Comment:
✨ Yes! Space complexity would actually be O(n) if the tree is unbalanced because of the recursive call stack. If the tree is balanced, space complexity would be O( log n)